### ISSUE TITLE
NTK update.

### ISSUE THREAD

**Issue by Stazed** [Stazed/ntk#1]
I have updated this "Unofficial" version of NTK for waf, the menu backspace bug,
and cleaned up the various compiler warnings. Most was simply C/P from FLTK 1.3.10.
Also some minor fixes that I had previously found.

Tested the build for non-mixer-xt (static) and non-timeline-xt, as well as zynaddsubfx.

This should be all that is needed to keep NTK running for the foreseeable future.
I will keep this fork updated here.

If @M0JXD or anyone else wants to update/maintain an "Official" version, you are welcome
to grab any of the changes here. Or, let me know and maybe I can push them to you...

Stazed

**Comment by M0JXD** [comment]
Thanks Stazed - I see you've been working some more in a WIP branch as well?

I've been slowly rolling in to this - I thought I'd have a whirl at bringing some maintenance to Non-Sequencer to expose me to NTK/FLTK. Instead I spent more time working on the build system and barely any time in the code, oh and I also seemed to break a lot of things. 😅

Still weighing up whether it's worth it. From an end user perspective, if FLTK added official/better support for overlapping widgets and allowed for transparency, and then the Vector and Cairo themes were ported to it would seem like everything NTK can do (from a visual perspective) would be doable in with FLTK. They already use a Cairo back-end for Wayland and by the sounds of it that will be back-ported for X11. You've already proven most of what's needed for Non is in FLTK already so that is probably a better effort. Of course it would be great to get the thoughts of other NTK projects too before diving in too deep.

Apparently there are some speed gains due to the NTK's architecture but in reality I think they'll be pretty marginal even on weak hardware like an early Raspberry Pi.

**Comment by Stazed** [comment]
Yeah, the wip branch, things are broken as well. Easy to do with this stuff...
Actually all that is needed to keep NTK going is just updating waf.  waf is actually included in some distros and is MIT license and maintained.  It should not violate any copyright if that were the only change...

I got the Vector and Cairo themes to build and work with FLTK (1.3.X) but... it causes all the popup menus to display wrong. Seems FLTK changed the drawing when using themes for the menu. So doubtful that they can be ported if they use cairo.
Maybe if someone were motivated, a similar looking theme could be designed without cairo.

Also FLTK is supposed to have cairo support, which needs to be enabled with the build, but I could not get it to work.

Gonna try build FLTK 1.4 and see what happens...

**Comment by M0JXD** [comment]
Yeah, that's all fair and good until a change is needed to wscript, which I think would be under LGPL like the rest of the library. 

The other nuisance point is that because waf compresses itself, and due to other things about how the project is run, it can be a pain to provide source packages for Debian et al.

Thanks for your work on this :)

**Comment by Stazed** [comment]
Got timeline to build with FLTK 1-4, but the Cairo and Vector themes still don't work.  Other display issues as well.
So kind of looks like keeping NTK alive might be the best option if you want that look...

**Comment by M0JXD** [comment]
The only theme I'd miss is Vector, not really for the transparency (that seems to have gone anyways) but because it's a flat theme that's rounded. Maybe a new theme based off on clean could be made to deliver that look in FLTK? There's a fair few themes made for FLTK Rust https://github.com/fltk-rs/fltk-theme that could be ported too ([I think some of them are ported for C++ here.](https://github.com/MoAlyousef/Fl_Theme))

@grammoboy2 as the other person keenly interested in NTK, what exactly do you miss when using the FLTK versions?

**Comment by grammoboy2** [comment]
Yes, I didn't saw the latest gui updates of JML as a improvement, I missed the (transparant?) look iirc. 

I use NTK, most of the time the original non-mixer. I only know the forked gui of non-session-manager (FLTK) looks quite ugly imho. No comparison with the original. 

JML did also some performance improvements iirc on NTK compared to FLTK indeed. At the end the question is how much this matters. JML saw his non-daw especially useful for RPI's iirc and they are also gained a lot more 'power' lately... 

How many projects are in a stable polished state in linuxaudio world? I would also take maintainability into considerations.

**Comment by M0JXD** [comment]
I'll have a look at reverting to the old Vector_Theme for NTK builds. Let me know what other ones you might like.

I think nsm-legacy-gui is themed with the pallete of Agordejo, which looks better under Qt Fusion (actually I wish Agordejo did not force it's theme and respected my GTK theming for Qt but that's a different issue.) The FLTK switch is understandable as it removes any requirement for them to try supporting a a whole GUI toolkit for what is, in their opinion, a legacy program that must be be maintained for back-compatibility.

I have an old Acer Aspire One, 1GHz Celeron and 2GB RAM running Mint XFCE, and a Raspberry Pi 4 4GB. Next chance I have I'll have a play with the Non stuff and see if there's any notable difference between NTK and FLTK on weak hardware.

**Comment by M0JXD** [comment]
I've added a new theme "Vector2" to [my repo](https://github.com/M0JXD/ntk) that restores the old vector look.

I've tested it and it makes Non-XT look just like the old Non as pictured [here](https://github.com/Stazed/non-timeline-xt/issues/3#issuecomment-2525171361).

Just reinstalling NTK should give the new theme unless you have the very latest Non-Mixer-XT which is now statically linked.

I can push the changes if you would like them @Stazed, although I understand you're still wary about maintaining an unofficial fork so no worries if not. As said I'll have more time to look at this next year.

Also: Merry Christmas 🎄

**Comment by Stazed** [comment]
Hey, the Vector2 theme does look nice!

I also added a couple themes to the FLTK build, a Vector (without cairo) and GTK from FLTK 1.3.
The Vector is a fair approximation of the NTK Vector. GTK looks pretty good.

I think I am going open a new repo with the title NTK-UNOFFICIAL and start from scratch on it.
Gonna put this repo to rest.

When I get the new repo up, I will grab the Vector2 theme and probably add the GTK one as well.

FYI, commit bb9f2f9f8dfebe75ab5d67f531b43e11a5805c21 was incorrect (my very first one) and I
grabbed the correct fix from FLTK 1.3 with e154789a04258c0c25b005208fcf8b16e09afcb9.

There may be some other broken things as well as I was just experimenting with things, so another reason to start with a clean repo.

It may be a while before I get to it...

**Comment by M0JXD** [comment]
I'm getting the feeling I've bullied you into supporting NTK 😬
Of course I'm really glad you're working on it but as I'm the person who said they'd contribute to this (at some point) please don't feel pressured to do so as well. I think for the (ten or so?) projects that use NTK switching to something that's the same API with a new name would be a non-issue (joke not intended).

The theme was created very quick and dirty, I just grabbed the old commit code and checked it worked 😅

**Comment by Stazed** [comment]
I guess the real issue for me is the difference between "official" vs "unofficial".
When you first posed the question of support for NTK, I was thinking "official" which I
do not want to tackle for various reasons. But an "unofficial" version should be as easy
as creating an NTK-UNOFFICIAL repo, dropping in a new waf, updating the top of the README
with a disclaimer and it's done. Supporting NTK unofficially in this way would require little
maintenance and is not a problem. I would rather support NTK "unofficially" than promise to
support some yet to be determined "official" NTK that I have no control over (no offense).

So after pondering this NTK support idea for a while, my actual preference would be to support NTK
"unofficially" and I think doing it in this way would comply with the GPL. It would also save you
the trouble of trying to revive an official version.

Would that work for you? It would for me...

I would have no problem adding the Vector2 theme as well as I want to add some other minor enhancements that are ABI compatible.

**Comment by M0JXD** [comment]
This sounds great! Honestly the more I thought about it the more appropriate it seemed to give such a project a new name even if that is just adding "unofficial"/"fork" afterwards.

If package maintainers want to start packaging your fork you could ask that's it's built as libntku.a/.so and then be symlinked against libntk.a/.so in a similar way to how Agordejo's nsm-legacy-gui symlinks against non-session-manager for backwards compatibility. For projects to switch properly instead of relying on a symlink would be very minor - usually just a couple of lines in the build system. There's only a few anyways.
